src/Compiler/CompilerUtils.cpp

```diff
@@ -307,8 +306,51 @@ struct Command {
 void setTargetCPU(const std::string &cpu) { mcpu = cpu; }
 void setTargetArch(const std::string &arch) { march = arch; }
 void setTargetTriple(const std::string &triple) { mtriple = triple; }
+void setOptLevel(const OptLevel level) { OptimizationLevel = level; }
 
-void LoadMLIR(string inputFilename, mlir::MLIRContext &context,
+static void setCompilerKeyValue(const OptionKind key, const string val) {
+  switch (key) {
+  case OptionKind::TargetTriple:
+    setTargetTriple(val);
+    return;
+  case OptionKind::TargetArch:
+    setTargetArch(val);
+    return;
+  case OptionKind::TargetCPU:
+    setTargetCPU(val);
+    return;
+  case OptionKind::CompilerOptLevel:
+    int level = (int)val[0] - (int)'0';
+    assert(level >= 0 && level <= 3 && "expected an OptLevel in [0..3] range");
+    setOptLevel((OptLevel)level);
+    return;
+  }
+  // In case there are options that were added but are unknown here, just ignore
+  // them.
+}
+
+// Set compiler context using a list of key/value pairs.
+void setCompileContext(mlir::MLIRContext &context,
+    const SmallVector<OptionKind, 4> key, const SmallVector<string, 4> val) {
```

Comment:
I prefer them too, but then I could not use the notation
```
setComplieContext({...}, {...});
```
notation because it required an explicit cast. So I chose to favor the implicit declarations (using SmallVector) rather than the more generic SmallVectorImpl.

Reply:
I changed the 2 lists into a list of pairs... no need for the hash function anyway. So that is clean and fast

Reply:
excellent